File: LBTAS.py
#!/usr/bin/env python3
"""
Leveson-Based Trade Assessment Scale (LBTAS)
============================================

A rigorous assessment methodology for digital commerce adapted from 
Nancy Leveson's Software Assessment Scale used in aircraft software development.

Copyright (C) 2024 Network Theory Applied Research Institute
Licensed under GNU Affero General Public License v3.0

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
"""

import json
import os
import sys
import argparse
from pathlib import Path
from typing import Dict, List, Optional, Union
from datetime import datetime

__version__ = "2.0.0"
__author__ = "Network Theory Applied Research Institute"
__license__ = "AGPL-3.0"

class LevesonRatingSystem:
    """
    Leveson-Based Trade Assessment Scale rating system.
    
    Provides a rigorous 6-point scale (-1 to +4) for evaluating
    digital commerce interactions based on Nancy Leveson's methodology.
    """
    
    DEFAULT_CATEGORIES = ['reliability', 'usability', 'performance', 'support']
    
    RATING_DESCRIPTIONS = {
        -1: "No Trust - User was harmed, exploited, or received a product or service with no discipline or malicious intent.",
        0: "Cynical Satisfaction - Interaction fulfills a basic promise requiring little to no discipline toward user satisfaction.",
        1: "Basic Promise - Interaction meets all articulated user demands, no more.",
        2: "Basic Satisfaction - Interaction meets socially acceptable standards exceeding articulated user demands.",
        3: "No Negative Consequences - Interaction designed to prevent loss, exceed basic quality.",
        4: "Delight - Interaction anticipates the evolution of user practices and concerns post-transaction."
    }

    # Display order for distributions: best (+4) down to worst (-1).
    RATING_LEVELS = [4, 3, 2, 1, 0, -1]

    # Short labels for each level (the part before " - " in RATING_DESCRIPTIONS).
    RATING_LABELS = {
        4: "Delight",
        3: "No Negative Consequences",
        2: "Basic Satisfaction",
        1: "Basic Promise",
        0: "Cynical Satisfaction",
        -1: "No Trust",
    }

    @staticmethod
    def _distribution(ratings: List[int]) -> Dict[str, int]:
        """Count ratings at each level. Keys are level strings -1..+4.

        Ratings are never averaged; the distribution (count per level) is the
        unit of reputation. See CLAUDE.md.
        """
        dist = {str(level): 0 for level in (-1, 0, 1, 2, 3, 4)}
        for rating in ratings:
            dist[str(rating)] += 1
        return dist

    @classmethod
    def _format_distribution(cls, dist: Dict[str, int], indent: str = "  ") -> str:
        """Render a distribution best-to-worst, one line per level: level label : count."""
        lines = []
        for level in cls.RATING_LEVELS:
            sign = f"+{level}" if level > 0 else (" 0" if level == 0 else f"{level}")
            label = cls.RATING_LABELS[level]
            lines.append(f"{indent}{sign} {label:<24}: {dist[str(level)]}")
        return "\n".join(lines)

    def __init__(self, storage_file: Optional[str] = None, categories: Optional[List[str]] = None):
        """
        Initialize the LBTAS rating system.
        
        Args:
            storage_file: Optional path to JSON file for persistent storage
            categories: Custom rating categories (defaults to reliability, usability, performance, support)
        """
        self.categories = categories or self.DEFAULT_CATEGORIES.copy()
        self.storage_file = storage_file
        self.exchanges = {}
        
        if self.storage_file:
            self.load_from_file()
    
    def add_exchange(self, name: str) -> None:
        """Add a new exchange to the system."""
        if name in self.exchanges:
            raise ValueError(f"Exchange '{name}' already exists.")
        
        self.exchanges[name] = {
            category: [] for category in self.categories
        }
        self.exchanges[name]['_metadata'] = {
            'created': datetime.now().isoformat(),
            'total_ratings': 0
        }
        
        if self.storage_file:
            self.save_to_file()
    
    def add_rating(self, exchange_name: str, criterion: str, rating: int) -> None:
        """
        Add a rating programmatically (non-interactive).
        
        Args:
            exchange_name: Name of the exchange
            criterion: Rating category
            rating: Rating value (-1 to 4)
        """
        if exchange_name not in self.exchanges:
            raise ValueError(f"Exchange '{exchange_name}' does not exist.")
        
        if criterion not in self.categories:
            raise ValueError(f"Criterion '{criterion}' not in valid categories: {self.categories}")
        
        if not isinstance(rating, int) or rating < -1 or rating > 4:
            raise ValueError(f"Rating must be integer between -1 and 4, got {rating}")
        
        self.exchanges[exchange_name][criterion].append(rating)
        self.exchanges[exchange_name]['_metadata']['total_ratings'] += 1
        
        if self.storage_file:
            self.save_to_file()
    
    def get_rating(self, criterion: str) -> int:
        """Get a rating from the user for a specific criterion (interactive)."""
        print(f"\nRate {criterion.capitalize()}:")
        print("=" * 50)
        
        for rating, description in self.RATING_DESCRIPTIONS.items():
            print(f" {rating:2d}: {description}")
        
        print("=" * 50)
        
        while True:
            try:
                rating = int(input(f"Enter your rating for {criterion.capitalize()} (-1 to 4): "))
                if -1 <= rating <= 4:
                    return rating
                else:
                    print("Please enter a rating between -1 and 4.")
            except ValueError:
                print("Invalid input. Please enter a number between -1 and 4.")
            except KeyboardInterrupt:
                print("\nRating cancelled.")
                sys.exit(0)
    
    def rate_exchange(self, name: str) -> None:
        """Rate an exchange based on Leveson Software Assessment Scale (interactive)."""
        if name not in self.exchanges:
            raise ValueError(f"Exchange '{name}' does not exist.")
        
        print(f"\nRating '{name}' using Leveson-Based Trade Assessment Scale")
        print("=" * 60)
        
        for criterion in self.categories:
            rating = self.get_rating(criterion)
            self.exchanges[name][criterion].append(rating)
            self.exchanges[name]['_metadata']['total_ratings'] += 1
        
        if self.storage_file:
            self.save_to_file()
        
        print(f"\nRating completed for '{name}'!")
    
    def view_ratings(self, name: str) -> Dict[str, Dict]:
        """Return the rating distribution for each criterion on an exchange.

        Ratings are never averaged. Each criterion maps to the count at every
        level (-1..+4) plus the category total::

            {"<category>": {"distribution": {"-1": 0, ..., "4": 0}, "total": 0}}
        """
        if name not in self.exchanges:
            raise ValueError(f"Exchange '{name}' does not exist.")

        summary = {}
        for criterion in self.categories:
            ratings = self.exchanges[name][criterion]
            summary[criterion] = {
                'distribution': self._distribution(ratings),
                'total': len(ratings),
            }
        return summary
    
    def get_all_exchanges(self) -> List[str]:
        """Return a list of all exchanges."""
        return list(self.exchanges.keys())
    
    def generate_report(self) -> Dict:
        """Generate a system report as rating distributions (never averages).

        Reputation is the count at each level plus the total. There is no
        system average, no category means, and no average-ranked performer
        lists. ``harm_flagged`` lists exchanges that received one or more -1
        ("No Trust") ratings, sorted by -1 count descending.
        """
        total_exchanges = len(self.exchanges)

        if total_exchanges == 0:
            return {
                'total_exchanges': 0,
                'total_ratings': 0,
                'overall_distribution': self._distribution([]),
                'category_distributions': {cat: self._distribution([]) for cat in self.categories},
                'exchange_distributions': {},
                'harm_flagged': [],
                'generated_at': datetime.now().isoformat(),
            }

        overall_ratings = []
        category_totals = {cat: [] for cat in self.categories}
        exchange_distributions = {}
        harm_counts = {}

        for exchange_name, exchange_data in self.exchanges.items():
            exchange_ratings = []
            for category in self.categories:
                ratings = exchange_data[category]
                category_totals[category].extend(ratings)
                exchange_ratings.extend(ratings)
                overall_ratings.extend(ratings)

            dist = self._distribution(exchange_ratings)
            exchange_distributions[exchange_name] = {
                'distribution': dist,
                'total': len(exchange_ratings),
            }
            if dist['-1'] > 0:
                harm_counts[exchange_name] = dist['-1']

        # Sort harm-flagged exchanges by -1 count descending, then name for stability.
        harm_flagged = sorted(harm_counts.items(), key=lambda kv: (-kv[1], kv[0]))

        return {
            'total_exchanges': total_exchanges,
            'total_ratings': len(overall_ratings),
            'overall_distribution': self._distribution(overall_ratings),
            'category_distributions': {
                cat: self._distribution(ratings) for cat, ratings in category_totals.items()
            },
            'exchange_distributions': exchange_distributions,
            'harm_flagged': [[name, count] for name, count in harm_flagged],
            'generated_at': datetime.now().isoformat(),
        }
    
    def save_to_file(self) -> None:
        """Save exchanges to JSON file."""
        if not self.storage_file:
            return
        
        try:
            with open(self.storage_file, 'w', encoding='utf-8') as f:
                json.dump(self.exchanges, f, indent=2, ensure_ascii=False)
        except Exception as e:
            print(f"Warning: Could not save to {self.storage_file}: {e}")
    
    def load_from_file(self) -> None:
        """Load exchanges from JSON file."""
        if not self.storage_file or not os.path.exists(self.storage_file):
            return
        
        try:
            with open(self.storage_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                # Validate loaded data structure
                for exchange_name, exchange_data in data.items():
                    if not isinstance(exchange_data, dict):
                        continue
                    # Ensure all categories exist
                    for category in self.categories:
                        if category not in exchange_data:
                            exchange_data[category] = []
                    # Ensure metadata exists
                    if '_metadata' not in exchange_data:
                        exchange_data['_metadata'] = {
                            'created': datetime.now().isoformat(),
                            'total_ratings': sum(len(ratings) for cat, ratings in exchange_data.items() if cat != '_metadata')
                        }
                
                self.exchanges = data
        except Exception as e:
            print(f"Warning: Could not load from {self.storage_file}: {e}")
    
    def export_to_csv(self, filename: str) -> None:
        """Export ratings to CSV (columns: exchange, category, rating, index).

        No timestamp column: the CLI store does not record per-rating times, so
        writing ``datetime.now()`` on every row would falsely imply each rating
        was made at export time. Honest per-rating timestamps live only in the
        API event records. ``index`` is the 1-based position of the rating
        within its category, matching the Go/Rust/TypeScript exports.
        """
        import csv

        with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
            fieldnames = ['exchange', 'category', 'rating', 'index']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

            writer.writeheader()
            for exchange_name, exchange_data in self.exchanges.items():
                for category in self.categories:
                    for index, rating in enumerate(exchange_data[category], start=1):
                        writer.writerow({
                            'exchange': exchange_name,
                            'category': category,
                            'rating': rating,
                            'index': index,
                        })

def main():
    """Command-line interface for LBTAS."""
    parser = argparse.ArgumentParser(
        description="Leveson-Based Trade Assessment Scale (LBTAS)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  lbtas rate --exchange "MyService"
  lbtas add --exchange "MyService" --criterion reliability --rating 3
  lbtas view --exchange "MyService"
  lbtas report
  lbtas export --format csv --output ratings.csv
        """
    )
    
    parser.add_argument('--version', action='version', version=f'LBTAS {__version__}')
    parser.add_argument('--storage', default='lbtas_ratings.json', 
                       help='Storage file for ratings (default: lbtas_ratings.json)')
    
    subparsers = parser.add_subparsers(dest='command', help='Available commands')
    
    # Rate command (interactive)
    rate_parser = subparsers.add_parser('rate', help='Rate an exchange interactively')
    rate_parser.add_argument('--exchange', required=True, help='Exchange name to rate')
    
    # Add command (programmatic)
    add_parser = subparsers.add_parser('add', help='Add a rating programmatically')
    add_parser.add_argument('--exchange', required=True, help='Exchange name')
    add_parser.add_argument('--criterion', required=True, help='Rating criterion')
    add_parser.add_argument('--rating', type=int, required=True, help='Rating value (-1 to 4)')
    
    # View command
    view_parser = subparsers.add_parser('view', help='View ratings for an exchange')
    view_parser.add_argument('--exchange', required=True, help='Exchange name to view')
    
    # List command
    list_parser = subparsers.add_parser('list', help='List all exchanges')
    
    # Report command
    report_parser = subparsers.add_parser('report', help='Generate system report')
    
    # Export command
    export_parser = subparsers.add_parser('export', help='Export ratings')
    export_parser.add_argument('--format', choices=['json', 'csv'], default='json', help='Export format')
    export_parser.add_argument('--output', required=True, help='Output filename')
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    # Initialize rating system
    rating_system = LevesonRatingSystem(storage_file=args.storage)
    
    try:
        if args.command == 'rate':
            if args.exchange not in rating_system.get_all_exchanges():
                rating_system.add_exchange(args.exchange)
            rating_system.rate_exchange(args.exchange)
            
        elif args.command == 'add':
            if args.exchange not in rating_system.get_all_exchanges():
                rating_system.add_exchange(args.exchange)
            rating_system.add_rating(args.exchange, args.criterion, args.rating)
            print(f"Added rating {args.rating} for {args.criterion} to {args.exchange}")
            
        elif args.command == 'view':
            ratings = rating_system.view_ratings(args.exchange)
            metadata = rating_system.exchanges[args.exchange].get('_metadata', {})
            created = metadata.get('created', 'unknown')
            exchange_total = sum(block['total'] for block in ratings.values())
            print(f"\nRatings for '{args.exchange}':")
            print("=" * 40)
            print(f"In service since: {created}")
            print(f"Total ratings (transaction volume): {exchange_total}")
            for criterion, block in ratings.items():
                print(f"\n{criterion.capitalize()} (total: {block['total']}):")
                print(LevesonRatingSystem._format_distribution(block['distribution']))

        elif args.command == 'list':
            exchanges = rating_system.get_all_exchanges()
            if exchanges:
                print("Registered exchanges:")
                for exchange in exchanges:
                    ratings = rating_system.view_ratings(exchange)
                    total = sum(block['total'] for block in ratings.values())
                    harm = sum(block['distribution']['-1'] for block in ratings.values())
                    line = f"  {exchange} ({total} ratings)"
                    if harm > 0:
                        line += f", {harm}x -1 No Trust"
                    print(line)
            else:
                print("No exchanges registered.")

        elif args.command == 'report':
            report = rating_system.generate_report()
            print("\nLBTAS System Report")
            print("=" * 50)
            print(f"Total exchanges: {report['total_exchanges']}")
            print(f"Total ratings (transaction volume): {report['total_ratings']}")

            print("\nOverall distribution:")
            print(LevesonRatingSystem._format_distribution(report['overall_distribution']))

            print("\nCategory distributions:")
            for category, dist in report['category_distributions'].items():
                print(f"  {category.capitalize()}:")
                print(LevesonRatingSystem._format_distribution(dist, indent="    "))

            if report['exchange_distributions']:
                print("\nPer-exchange distributions:")
                for exchange_name, block in report['exchange_distributions'].items():
                    created = rating_system.exchanges[exchange_name].get('_metadata', {}).get('created', 'unknown')
                    print(f"  {exchange_name} (transaction volume: {block['total']}, in service since: {created}):")
                    print(LevesonRatingSystem._format_distribution(block['distribution'], indent="    "))

            if report['harm_flagged']:
                print("\nHarm-flagged exchanges (received -1 No Trust):")
                for exchange_name, count in report['harm_flagged']:
                    print(f"  {exchange_name}: {count}x -1")
                    
        elif args.command == 'export':
            if args.format == 'json':
                with open(args.output, 'w') as f:
                    json.dump(rating_system.exchanges, f, indent=2)
            elif args.format == 'csv':
                rating_system.export_to_csv(args.output)
            print(f"Exported to {args.output}")
            
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        sys.exit(0)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

# Example usage when run as script
if __name__ == "__main__":
    if len(sys.argv) == 1:
        # Interactive demo mode
        print("LBTAS Interactive Demo")
        print("=" * 50)
        print("Welcome to the Leveson-Based Trade Assessment Scale!")
        print("This demo will walk you through rating a service.\n")
        
        rating_system = LevesonRatingSystem()
        
        try:
            # Demo exchange
            demo_exchange = "DemoService"
            print(f"Creating demo exchange: '{demo_exchange}'")
            rating_system.add_exchange(demo_exchange)
            
            print("\nYou can now rate this service on four criteria:")
            print("- Reliability: How dependable and consistent is it?")
            print("- Usability: How easy and pleasant is it to use?") 
            print("- Performance: How fast and efficient is it?")
            print("- Support: How helpful is customer service?")
            
            print("\nWould you like to rate this demo service? (y/n): ", end="")
            if input().lower().startswith('y'):
                rating_system.rate_exchange(demo_exchange)
                
                # Show results as a distribution (ratings are never averaged).
                ratings = rating_system.view_ratings(demo_exchange)
                print(f"\nYour ratings for '{demo_exchange}':")
                print("-" * 30)
                for criterion, block in ratings.items():
                    print(f"\n{criterion.capitalize()} (total: {block['total']}):")
                    print(LevesonRatingSystem._format_distribution(block['distribution']))
                print("\nRatings are recorded as a distribution (count per level), never averaged.")
            else:
                print("Demo completed. Try 'python lbtas.py --help' for CLI options.")
                
        except KeyboardInterrupt:
            print("\nDemo cancelled. Goodbye!")
        except Exception as e:
            print(f"Demo error: {e}")
    else:
        # Use CLI interface
        main()

File: lbtas.ts
/**
 * Leveson-Based Trade Assessment Scale (LBTAS)
 * 
 * A rating system for digital commerce based on Nancy Leveson's
 * aircraft software assessment methodology.
 * 
 * Copyright (C) 2024 Network Theory Applied Research Institute
 * Licensed under GNU Affero General Public License v3.0
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

import * as fs from 'fs';
import * as readline from 'readline';

const VERSION = '2.0.0';
const AUTHOR = 'Network Theory Applied Research Institute';
const LICENSE = 'AGPL-3.0';

interface ExchangeData {
  [category: string]: number[];
  _metadata: {
    created: string;
    total_ratings: number;
  };
}

interface StorageData {
  [exchange: string]: ExchangeData;
}

// A distribution is the count of ratings at each level, keyed by level string
// ("-1".."4"). Ratings are never averaged; this is the unit of reputation.
type Distribution = { [level: string]: number };

interface CategorySummary {
  distribution: Distribution;
  total: number;
}

interface RatingSummary {
  [category: string]: CategorySummary;
}

interface SystemReport {
  total_exchanges: number;
  total_ratings: number;
  overall_distribution: Distribution;
  category_distributions: { [category: string]: Distribution };
  exchange_distributions: { [exchange: string]: CategorySummary };
  // [exchange, -1 count] for exchanges with any -1, sorted by count descending.
  harm_flagged: Array<[string, number]>;
  generated_at: string;
}

// Display order for distributions: best (+4) to worst (-1).
const RATING_LEVELS = [4, 3, 2, 1, 0, -1];

// Short label for each rating level.
const RATING_LABELS: { [key: string]: string } = {
  '4': 'Delight',
  '3': 'No Negative Consequences',
  '2': 'Basic Satisfaction',
  '1': 'Basic Promise',
  '0': 'Cynical Satisfaction',
  '-1': 'No Trust',
};

function newDistribution(): Distribution {
  return { '-1': 0, '0': 0, '1': 0, '2': 0, '3': 0, '4': 0 };
}

function distributionOf(ratings: number[]): Distribution {
  const dist = newDistribution();
  for (const rating of ratings) {
    const key = String(rating);
    dist[key] = (dist[key] || 0) + 1;
  }
  return dist;
}

// Render a distribution best-to-worst, one line per level: level label : count.
function formatDistribution(dist: Distribution, indent: string = '  '): string {
  const lines: string[] = [];
  for (const level of RATING_LEVELS) {
    const sign = level > 0 ? `+${level}` : (level === 0 ? ' 0' : `${level}`);
    const label = RATING_LABELS[String(level)];
    lines.push(`${indent}${sign} ${label.padEnd(24)}: ${dist[String(level)]}`);
  }
  return lines.join('\n');
}

class LevesonRatingSystem {
  private readonly DEFAULT_CATEGORIES = ['reliability', 'usability', 'performance', 'support'];
  
  private readonly RATING_DESCRIPTIONS: { [key: number]: string } = {
    '-1': 'No Trust - User was harmed, exploited, or received a product or service with no discipline or malicious intent.',
    '0': 'Cynical Satisfaction - Interaction fulfills a basic promise requiring little to no discipline toward user satisfaction.',
    '1': 'Basic Promise - Interaction meets all articulated user demands, no more.',
    '2': 'Basic Satisfaction - Interaction meets socially acceptable standards exceeding articulated user demands.',
    '3': 'No Negative Consequences - Interaction designed to prevent loss, exceed basic quality.',
    '4': 'Delight - Interaction anticipates the evolution of user practices and concerns post-transaction.'
  };

  private categories: string[];
  private storagePath: string | null;
  private exchanges: StorageData;

  constructor(storagePath: string | null = null, categories: string[] | null = null) {
    this.categories = categories || this.DEFAULT_CATEGORIES.slice();
    this.storagePath = storagePath;
    this.exchanges = {};

    if (this.storagePath) {
      this.loadFromFile();
    }
  }

  private loadFromFile(): void {
    if (!this.storagePath) return;
    
    try {
      if (fs.existsSync(this.storagePath)) {
        const data = fs.readFileSync(this.storagePath, 'utf8');
        this.exchanges = JSON.parse(data);
      }
    } catch (error) {
      console.error(`Error loading from file: ${error}`);
    }
  }

  private saveToFile(): void {
    if (!this.storagePath) return;
    
    try {
      fs.writeFileSync(this.storagePath, JSON.stringify(this.exchanges, null, 2));
    } catch (error) {
      console.error(`Error saving to file: ${error}`);
    }
  }

  addExchange(name: string): void {
    if (this.exchanges[name]) {
      throw new Error(`Exchange '${name}' already exists.`);
    }

    const exchangeData: ExchangeData = {
      _metadata: {
        created: new Date().toISOString(),
        total_ratings: 0
      }
    };

    for (const category of this.categories) {
      exchangeData[category] = [];
    }

    this.exchanges[name] = exchangeData;
    this.saveToFile();
  }

  addRating(exchangeName: string, criterion: string, rating: number): void {
    if (!this.exchanges[exchangeName]) {
      throw new Error(`Exchange '${exchangeName}' does not exist.`);
    }

    if (!this.categories.includes(criterion)) {
      throw new Error(`Criterion '${criterion}' not in valid categories: ${this.categories.join(', ')}`);
    }

    if (!Number.isInteger(rating) || rating < -1 || rating > 4) {
      throw new Error(`Rating must be integer between -1 and 4, got ${rating}`);
    }

    this.exchanges[exchangeName][criterion].push(rating);
    this.exchanges[exchangeName]._metadata.total_ratings++;
    this.saveToFile();
  }

  async getRating(criterion: string): Promise<number> {
    console.log(`\nRate ${criterion.charAt(0).toUpperCase() + criterion.slice(1)}:`);
    console.log('='.repeat(50));
    
    for (const [rating, description] of Object.entries(this.RATING_DESCRIPTIONS)) {
      console.log(`${rating.padStart(2)}: ${description}`);
    }
    
    console.log('='.repeat(50));

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    return new Promise((resolve) => {
      const askRating = () => {
        rl.question(`Enter your rating for ${criterion.charAt(0).toUpperCase() + criterion.slice(1)} (-1 to 4): `, (answer) => {
          const rating = parseInt(answer);
          if (!isNaN(rating) && rating >= -1 && rating <= 4) {
            rl.close();
            resolve(rating);
          } else {
            console.log('Please enter a rating between -1 and 4.');
            askRating();
          }
        });
      };
      askRating();
    });
  }

  async rateExchange(name: string): Promise<void> {
    if (!this.exchanges[name]) {
      throw new Error(`Exchange '${name}' does not exist.`);
    }

    console.log(`\nRating '${name}' using Leveson-Based Trade Assessment Scale`);
    console.log('='.repeat(60));

    for (const criterion of this.categories) {
      const rating = await this.getRating(criterion);
      this.exchanges[name][criterion].push(rating);
      this.exchanges[name]._metadata.total_ratings++;
    }

    this.saveToFile();
    console.log(`\nRating completed for '${name}'!`);
  }

  viewRatings(name: string): RatingSummary {
    if (!this.exchanges[name]) {
      throw new Error(`Exchange '${name}' does not exist.`);
    }

    const summary: RatingSummary = {};

    for (const criterion of this.categories) {
      const ratings = this.exchanges[name][criterion];
      summary[criterion] = {
        distribution: distributionOf(ratings),
        total: ratings.length,
      };
    }

    return summary;
  }

  getAllExchanges(): string[] {
    return Object.keys(this.exchanges);
  }

  // The per-exchange "in service since" date stored at creation time.
  getCreated(name: string): string {
    const exchange = this.exchanges[name];
    return exchange && exchange._metadata ? exchange._metadata.created : 'unknown';
  }

  generateReport(): SystemReport {
    const totalExchanges = Object.keys(this.exchanges).length;

    if (totalExchanges === 0) {
      const categoryDistributions: { [category: string]: Distribution } = {};
      for (const category of this.categories) {
        categoryDistributions[category] = newDistribution();
      }
      return {
        total_exchanges: 0,
        total_ratings: 0,
        overall_distribution: newDistribution(),
        category_distributions: categoryDistributions,
        exchange_distributions: {},
        harm_flagged: [],
        generated_at: new Date().toISOString()
      };
    }

    const overall: number[] = [];
    const categoryTotals: { [key: string]: number[] } = {};
    const exchangeDistributions: { [exchange: string]: CategorySummary } = {};
    const harmFlagged: Array<[string, number]> = [];

    for (const category of this.categories) {
      categoryTotals[category] = [];
    }

    for (const [exchangeName, exchangeData] of Object.entries(this.exchanges)) {
      const exchangeRatings: number[] = [];

      for (const category of this.categories) {
        const ratings = exchangeData[category];
        for (const r of ratings) {
          categoryTotals[category].push(r);
          exchangeRatings.push(r);
          overall.push(r);
        }
      }

      const dist = distributionOf(exchangeRatings);
      exchangeDistributions[exchangeName] = {
        distribution: dist,
        total: exchangeRatings.length
      };
      if (dist['-1'] > 0) {
        harmFlagged.push([exchangeName, dist['-1']]);
      }
    }

    const categoryDistributions: { [category: string]: Distribution } = {};
    for (const category of this.categories) {
      categoryDistributions[category] = distributionOf(categoryTotals[category]);
    }

    // Sort harm-flagged by -1 count descending, then name ascending for stability.
    harmFlagged.sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));

    return {
      total_exchanges: totalExchanges,
      total_ratings: overall.length,
      overall_distribution: distributionOf(overall),
      category_distributions: categoryDistributions,
      exchange_distributions: exchangeDistributions,
      harm_flagged: harmFlagged,
      generated_at: new Date().toISOString()
    };
  }

  exportToJSON(outputPath: string): void {
    fs.writeFileSync(outputPath, JSON.stringify(this.exchanges, null, 2));
  }

  exportToCSV(outputPath: string): void {
    // Columns: exchange, category, rating, index (1-based position within the
    // category). No timestamp: the CLI store does not record per-rating times,
    // so a column of export-time stamps would be dishonest. Honest per-rating
    // timestamps live only in the API event records.
    const lines: string[] = ['exchange,category,rating,index'];

    for (const [exchangeName, exchangeData] of Object.entries(this.exchanges)) {
      for (const category of this.categories) {
        const ratings = exchangeData[category];
        for (let i = 0; i < ratings.length; i++) {
          lines.push(`${exchangeName},${category},${ratings[i]},${i + 1}`);
        }
      }
    }

    fs.writeFileSync(outputPath, lines.join('\n'));
  }
}

// CLI interface
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('LBTAS - Leveson-Based Trade Assessment Scale');
    console.log('\nUsage:');
    console.log('  node lbtas.js rate <exchange>');
    console.log('  node lbtas.js add <exchange> <criterion> <rating>');
    console.log('  node lbtas.js view <exchange>');
    console.log('  node lbtas.js list');
    console.log('  node lbtas.js report');
    console.log('  node lbtas.js export <format> <output>');
    return;
  }

  const command = args[0];
  const storage = 'lbtas_ratings.json';
  const system = new LevesonRatingSystem(storage);

  try {
    switch (command) {
      case 'rate': {
        const exchange = args[1];
        if (!exchange) {
          console.error('Error: Exchange name required');
          return;
        }
        if (!system.getAllExchanges().includes(exchange)) {
          system.addExchange(exchange);
        }
        await system.rateExchange(exchange);
        break;
      }

      case 'add': {
        const [, exchange, criterion, ratingStr] = args;
        if (!exchange || !criterion || !ratingStr) {
          console.error('Error: exchange, criterion, and rating required');
          return;
        }
        const rating = parseInt(ratingStr);
        if (!system.getAllExchanges().includes(exchange)) {
          system.addExchange(exchange);
        }
        system.addRating(exchange, criterion, rating);
        console.log(`Added rating ${rating} for ${criterion} to ${exchange}`);
        break;
      }

      case 'view': {
        const exchange = args[1];
        if (!exchange) {
          console.error('Error: Exchange name required');
          return;
        }
        const ratings = system.viewRatings(exchange);
        const exchangeTotal = Object.values(ratings).reduce((sum, block) => sum + block.total, 0);
        console.log(`\nRatings for '${exchange}':`);
        console.log('='.repeat(40));
        console.log(`In service since: ${system.getCreated(exchange)}`);
        console.log(`Total ratings (transaction volume): ${exchangeTotal}`);
        for (const [criterion, block] of Object.entries(ratings)) {
          const name = criterion.charAt(0).toUpperCase() + criterion.slice(1);
          console.log(`\n${name} (total: ${block.total}):`);
          console.log(formatDistribution(block.distribution));
        }
        break;
      }

      case 'list': {
        const exchanges = system.getAllExchanges();
        if (exchanges.length === 0) {
          console.log('No exchanges registered.');
        } else {
          console.log('Registered exchanges:');
          for (const exchange of exchanges) {
            const ratings = system.viewRatings(exchange);
            const total = Object.values(ratings).reduce((s, block) => s + block.total, 0);
            const harm = Object.values(ratings).reduce((s, block) => s + block.distribution['-1'], 0);
            let line = `  ${exchange} (${total} ratings)`;
            if (harm > 0) {
              line += `, ${harm}x -1 No Trust`;
            }
            console.log(line);
          }
        }
        break;
      }

      case 'report': {
        const report = system.generateReport();
        console.log('\nLBTAS System Report');
        console.log('='.repeat(50));
        console.log(`Total exchanges: ${report.total_exchanges}`);
        console.log(`Total ratings (transaction volume): ${report.total_ratings}`);

        console.log('\nOverall distribution:');
        console.log(formatDistribution(report.overall_distribution));

        console.log('\nCategory distributions:');
        for (const [category, dist] of Object.entries(report.category_distributions)) {
          const name = category.charAt(0).toUpperCase() + category.slice(1);
          console.log(`  ${name}:`);
          console.log(formatDistribution(dist, '    '));
        }

        if (Object.keys(report.exchange_distributions).length > 0) {
          console.log('\nPer-exchange distributions:');
          for (const [exchangeName, block] of Object.entries(report.exchange_distributions)) {
            console.log(`  ${exchangeName} (transaction volume: ${block.total}, in service since: ${system.getCreated(exchangeName)}):`);
            console.log(formatDistribution(block.distribution, '    '));
          }
        }

        if (report.harm_flagged.length > 0) {
          console.log('\nHarm-flagged exchanges (received -1 No Trust):');
          for (const [exchangeName, count] of report.harm_flagged) {
            console.log(`  ${exchangeName}: ${count}x -1`);
          }
        }
        break;
      }

      case 'export': {
        const format = args[1];
        const output = args[2];
        if (!format || !output) {
          console.error('Error: format and output path required');
          return;
        }
        if (format === 'json') {
          system.exportToJSON(output);
        } else if (format === 'csv') {
          system.exportToCSV(output);
        } else {
          console.error('Error: format must be json or csv');
          return;
        }
        console.log(`Exported to ${output}`);
        break;
      }

      default:
        console.error(`Unknown command: ${command}`);
    }
  } catch (error) {
    console.error(`Error: ${error}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { LevesonRatingSystem };
